### ISSUE TITLE
Support automated reconnection (and other actions when the network connection drops)

### ISSUE THREAD

**Issue by MattBrittan** [eclipse/paho.golang#26]
**Is your feature request related to a problem? Please describe.**

When the network link with the MQTT broker drops there is currently no obvious way to detect or handle this. There is an `OnDisconnect` call back but this is only called when the broker cleanly disconnects (i.e. a Disconnect notification is received from the broker). I believe that the most common reason to want to detect the loss of connection is so the connection can be re-established so that is the use-case I am going to focus on (but any solution must provide a means of detecting loss of connection).

The lack of an "obvious way" for a user to implement autoreconnection functionality with the library as it currently stands does not mean that it's impossible; but it would be difficult. The challanges I see are:

* When Publish returns an error how do I tell if the message has been stored and will be retried when the connection returns (as the store is not yet implemented this is a bit academic).
* It would be difficult to determine if any errors returned are connection related or due to something else (you could ignore those and just use the keepalive but I don't think that is a good approach)
* If the connection has gone down and we want to reconnect then we will need a threadsafe way of doing so. It is easy to introduce bugs here because there may be many goroutines attempting to call `Publish`, the store needs to be cleanly shutdown/reopened and we need to bring up a new connection.
* The connection may be down for some time and we may want to use the inbuilt store rather than implementing our own form of persistence (I would also like to be able to call `Publish()` before `Connect()` and have the message added to the store).

Because automatic reconnection is a common requirement I don't think its appropriate to leave all of the above to the user. This project should provide a method and/or recommended approach.

Note: Some of how this works will depend upon how persistence is implemented. The MQTT [spec dictates](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901238) that there are situations where unacknowledged PUBLISH (and related messages) MUST be resent. How this resend is handled will have an impact on how reconnections can be processed.

*Apologies - I may have gone overboard on the detail in this issue! However I believe its important to get this right because it has a significant impact on the ease-of-use of the package*

**Describe the solution you'd like**

My proposed solution contains three elements:

1. A clear way of telling if an error can be safely ignored (because the message will be delivered when the connection comes up).
2. A means of detecting that the connection has been lost.
3. A clean way of re-establishing a connection (making some assumptions about the store).

***Error Handling***

Any errors returned by a function that has stored the relevant message (e.g. in `MemoryPersistence`) will be wrapped by a `MessagePersistedError` i.e.: 

```
type MessagePersistedError struct {
	error
}
```

This will make it fairly simple for users to ignore such messages if they have implemented a persistent store and some form of reconnection logic (meaning its fairly safe to assume they will be delivered eventually) i.e.

```
var e *MessagePersistedError
if err != nil && !errors.As(err, &e) {
	// This message will not be sent so we better do something (can generally ignore any MessagePersistedError)...
	fmt.Println("error!")
}
```

***Detect loss of connection***

In the v3 client a callback was used for this however I feel that using a channel is a cleaner approach more in keeping with the way `Context` works and redues the chance of races etc:

```
func (c *Client) Done() <-chan struct{} {
  return c.shutdown;
}
```

The returned channel will be closed when the client has completed shutting down. A shutdown will be automatically triggered by any a network failure (error returned by `Read` or `Write` on the `net.Conn`). 

Note: The below assumes that the store is not closed until `Disconnect` is called and that any calls to `Publish` will just add the packet to the store and return an error (which is what would happen now because the connection is unavailable).

***Clean way to reestablish connection***

While it would be possible to add the reconnection logic into `Client` I believe it is better to use a wrapper. Doing this means that users who do not want autoreconnection can ignore this and the reconnection logic is nicely isolated (providing an example for users wanting to take a different approach to this; e.g. try more than one broker).

There are a few issues with creating a new `Client` (the cleanest approach):

1. We need to be sure that the store is properly closed before the new client opens it
2. The user needs a way to transition between the two clients (ideally without needing to handle persistence themselves). Calling `Publish` while the connection is down would return an error but the message would have been added to the store (detectable via `MessagePersistedError`)

This can be implemented as a wrapper around the package; functions such as `Publish` will be wrapped ensuring they hold a read lock on an `RWMutex`:

```
func (rc *ReconnectClient) Subscribe(ctx context.Context, s *Subscribe) (*Suback, error) {
   rc.RLock()
   defer rc.RUnlock()
   return rc.c.Subscribe(ctc, s)
}
```

The reconnection logic will be something like the following (started as a go routine):

```
func (rc *ReconnectClient) watchConnection() {
	for {
		select {
		case <-rc.stop:
			debug.Println("watchConnection stopping")
			return
		case <-rc.c.Done():
			break
		}
		debug.Println("reconnection Starting")

		for {
			newConn := mqtt.NewClient()
			// Code to build a new connection (may include a callback to a user provided function to establish the connection)
			// Should keep an eye on `rc.stop` in case a shutdown is requested

			// Once we are ready to call Client.Connect()
			rc.WLock()
			_, err := newConn.Connect()
			if err == nil {
				rc.c.Disconnect() // Close the store in the existing connection
				rc.c = newConn    // Move over to the new connection
				rc.WUnlock()
				go rc.OnReconnect() // Callback so user can resubscribe or whatever
				break               // successfully reconnected
			}
			rc.WUnlock()
			// Maybe a pause before trying to connect again
		}
	}
}
```

Finally `Disconnect` would close `rc.stop` before calling the `rc.c.Disconnect()`.

**Describe alternatives you've considered**

***Active Requests***

I think its important to consider how archive requests (e.g. a call currently blocked at `Publish`) should be handled before looking at how to implement this. I see two options:

1. The connection is re-established and the request is retried 
2. In progress functions return an error when the connection is lost (but the persistence functionality takes care of resending relevant packets upon reconnection)

From a user perspective I would say that option 1 is generally preferable. However there is quite a bit of complexity around this including:

* How to handle things like `Subscribe` (its common to resubscribe to all relevant topics upon connection to ensure that the subscription is current so retrying such requests may not make sense)
* There may be many requests in progress at different stages of completion (e.g. with a publish has PUBCOMP/PUBREC been received; has the reply been sent?)
* As mentioned above there are cases where the MQTT spec requires that we resend some packets when a new connection is established.
* How do we handle calls to (i.e.) `Publish` made while the connection is down (taking into account that the outage could be long). Does this depend upon the QOS level (do we just drop QOS 0 requests and block on others etc).
* How do we detect if the store does not actually store anything (e.g. `noopPersistence`)?

So my feeling is that the complexity of option 1 outweighs the benefits; I believe that most users will be happy so long as they can confirm that their `Publish` has been stored and will be delivered eventually (knowing exactly when it has arrived at the broker is fairly meaningless because there may be a delay between the broker and subscriber).

***alternatives***

The simplest approach would have been to add a `ConnectionLostError` wrapper around errors that related to a lost connection. This would make it easier for the end user to detect that the connection was lost (but leaves the full responsibility of dealing with it to them). Note that many users would need to use the Pinger because its common to subscribe to a few topics and then just wait for messages meaning that after the initial connection setup no further calls to library functions are made.

Adding a simple `OnConnectionLost()` callback, or perhaps having connect return a channel that is closed when the library shuts down are other options. This provides a simpler method to detect the loss of connection but leaves all of the work of reestablishing the connection to the user.

The next alternative considered was a more complicated callback; something along the lines of:

```
// Connection represents an initialised MQTT connection (i.e. the handshake has been successfully completed)
type Connection interface {
   Conn() net.Conn // Returns the underlying connection

   // It may be neccessary to Authenticate or reestablish subscriptions after the connection is re-established
   // The function passed in here will be called in a go-routine as soon as the new client is active
   SetLiveCallback(func(ctx context.Context, c *Client))
}

// Connector provides the ability to initialise an MQTT connection (completing the handshake etc) without 
// building a new Client (this is used in the OnConnectionError callback)
type Connector interface {
   Connect(ctx context.Context, cp *Connect) (Connection, *Connack, error))
}

type Client struct {
   ...
   // OnConnectionError will be called if an operation on the connection (net.Conn) returns an error (the err parameter)
   // The callback may use the passed in error to determine what action to take and:
   //  - Establish a new connection, use `Connect` to complete the required MQTT handshakes, and return the new Connection
   //  - Determine that the operation should be re (i.e. the connection is viable and the operation should be retried) and return thc conn 
it was passed.
   //  - Return `nil` which signifies that no further action should be taken
   // This callback will not be called if the disconnection is due to, or coincides with, a call to `Disconnect()`
   // Note: As per the MQTT spec any retained messages will be retried if a new connection is established
   OnConnectionError func (ctx context.Context, err error.Error, conn Connection, connector lostConn) Connection
}
```

While I was initially keen on this approach there were too many questions around how to handle calls that were in progress (may I suggest adding some form of `token`  :-( ).

**Additional context**
Some discussion on issue #23.

**Comment by MattBrittan** [comment]
I have had a go at writing a wrapper that handles connection/reconnection (and includes an example running under docker); currently this is [in my repo](https://github.com/ChIoT-Tech/paho.golang/tree/master/autopaho). Happy to submit it here or publish as a separate project. I encountered a couple of issues (possibly just re documentation) whilst implementing this:

* I have had to assume that any call to OnClientError is fatal (and the connection will be dropped). I believe this is currently the case but don't think it's documented anywhere? (also I believe that OnClientError may be called multiple times so users need to account for that)
* There does not appear to be any way to wait for the paho client to shutdown (you have the workers WaitGroup but nothing ever waits on this). I don't think this is a huge issue currently but is likely to be more of an issue when persistence support is added.

**Comment by abh** [comment]
With autopaho merged into the repository, it'd be helpful for new users to the library if this issue was updated (or superseded) with whatever the current status is.

**Comment by MattBrittan** [comment]
Closing this issue because [auto Paho](https://github.com/eclipse/paho.golang/tree/master/autopaho) ([docs](https://pkg.go.dev/github.com/eclipse/paho.golang/autopaho)) now provides a wrapper around the library which implements auto reconnection (providing a solution to the issue as well as an example for others who wish to implement this themselves.
